File: pagerank/page_rank.py
from pyspark import SparkContext, SparkConf
import re
import sys

DAMPING_FACTOR = 0.8


def data_parser(line):
    # get the index of the begin of the title
    begin_title_index = line.find("<title>") + len("<title>")
    # get the index of the end of the title
    end_title_index = line.find("</title>")
    # get the title
    page_title = line[begin_title_index:end_title_index]

    # get the index of the begin of the text section
    begin_text_index = line.find("<text") + len("<text")
    # get the index of the end of the text section
    end_text_index = line.find("</text>")
    # get the text section
    page_text_section = line[begin_text_index:end_text_index]
    # get all the outgoing_links (any character between 2 pairs of '[]'
    outgoing_links = re.findall(r'\[\[([^]]*)\]\]', page_text_section)

    return page_title, outgoing_links


def spread_rank(node, outgoing_links, rank):
    # the contribution of a node to itself is 0
    rank_list = [(node, 0)]
    if len(outgoing_links) > 0:
        mass_to_send = rank / len(outgoing_links)
        for link in outgoing_links:
            rank_list.append((link, mass_to_send))
    return rank_list


if __name__ == "__main__":
    if len(sys.argv) != 4:
        # args: [1]=input_file(txt),   [2]=output_file(txt),    [3]=nIter
        print("Usage: page_rank.py <input_file> <output> <iterations>", file=sys.stderr)
        sys.exit(-1)

    # import context from Spark (distributed computing using yarn, name of the application)
    conf = SparkConf().setMaster("yarn").setAppName("page_rank_baggins").set("spark.dynamicAllocation.enabled", True)
    sc = SparkContext(conf=conf)

    # import input data from txt file to rdd
    input_data_rdd = sc.textFile(sys.argv[1])

    DAMPING_FACTOR_BR = sc.broadcast(DAMPING_FACTOR)

    # count number of nodes in the input dataset, broadcast the value (equal for each worker)
    node_number = input_data_rdd.count()
    node_number_br = sc.broadcast(node_number)

    # parse input rdd to get graph structure (k=title, v=[outgoing links])
    nodes = input_data_rdd.map(lambda input_line: data_parser(input_line)).cache()

    considered_keys = nodes.keys().collect()

    # set the initial pagerank (1/node_number)
    page_ranks = nodes.mapValues(lambda value: 1 / node_number_br.value)

    for i in range(int(sys.argv[3])):
        # computes masses to send (node_tuple[0] = title | node_tuple[1][0] = outgoing_links | node_tuple[1][1] = rank)
        contribution_list = nodes.join(page_ranks) \
            .flatMap(lambda node_tuple: spread_rank(node_tuple[0], node_tuple[1][0], node_tuple[1][1]))

        # inner join to consider only nodes inside the considered network
        considered_contributions = contribution_list.filter(lambda x: x[0] in considered_keys)

        # aggregate contributions for each node, compute final ranks
        page_ranks = considered_contributions.reduceByKey(lambda x, y: x + y) \
            .mapValues(lambda summed_contributions: (float(1 - DAMPING_FACTOR_BR.value) / node_number) +
                                                    (DAMPING_FACTOR_BR.value * float(summed_contributions)))

    # swap key and value, sort by key (by pagerank) and swap again
    sorted_page_ranks = page_ranks.map(lambda a: (a[1], a[0])) \
        .sortByKey(False) \
        .map(lambda a: (a[1], a[0]))

    # save the output
    sorted_page_ranks.saveAsTextFile(sys.argv[2])

File: pagerank_Java/src/main/java/PageRank.java
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.broadcast.Broadcast;
import scala.Tuple2;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class PageRank {
    private static final double DUMPING_FACTOR = 0.8;

    public static void main(String[] args) {
        // import context from Spark (distributed computing using yarn, set name of the application)
        SparkConf sparkConf = new SparkConf().setAppName("pageRankJava").setMaster("yarn");
        JavaSparkContext javaSparkContext = new JavaSparkContext(sparkConf);

        // import input data from txt file to rdd
        JavaRDD<String> inputDataRDD = javaSparkContext.textFile(args[0]);

        // count number of nodes in the input dataset (the N number)
        long nodesNumber = inputDataRDD.count();

        // parse input rdd to get graph structure (k=title, v=[outgoing links])
        // the result is cached since it is static and to be accessible by every worker
        JavaPairRDD<String, ArrayList<String>> nodesRDD = inputDataRDD.mapToPair(DataParser::parseData).cache();

        // set the initial pagerank (1/node_number)
        JavaPairRDD<String, Double> pageRankRDD = nodesRDD.mapValues(value -> 1 / (double) nodesNumber);

        // list of title of the considered nodes. Explicitly computed to avoid the additional join
        List<String> consideredKeys = nodesRDD.keys().collect();

        for(int i = 0; i< Integer.parseInt(args[2]); i++) {
            // computes masses to send (node_tuple._1 = title | node_tuple._2._1 = outgoing_links | node_tuple._2._2 = rank)
            JavaPairRDD<String, Double> contributionRDD = nodesRDD.join(pageRankRDD).flatMapToPair(node_tuple -> spreadRank(node_tuple._1, node_tuple._2._1, node_tuple._2._2));

            // take the only contributions that are relative to considered nodes
            JavaPairRDD<String, Double> consideredContributionsRDD = contributionRDD.filter(x -> consideredKeys.contains(x._1));

            // aggregate contributions for each node, compute final ranks
            pageRankRDD = consideredContributionsRDD.reduceByKey(Double::sum)
                    .mapValues(summedContributions -> (1 - DUMPING_FACTOR) / nodesNumber + DUMPING_FACTOR * summedContributions);
        }

        // sort by value (pagerank)
        JavaRDD<Tuple2<String, Double>> sortedPageRankRDD = pageRankRDD.map(pageRank -> pageRank).sortBy(page -> page._2, false, 1);
        // save the output
        sortedPageRankRDD.saveAsTextFile(args[1]);
    }

    private static Iterator<Tuple2<String, Double>> spreadRank(String title, ArrayList<String> outgoingLinks, Double pageRank) {
        ArrayList<Tuple2<String, Double>> result = new ArrayList<>();
        // the contribution of a node to itself is 0
        result.add(new Tuple2<>(title, 0.0));
        // for any other outgoing link, this nodes contributes by rank/N, so we save (link-contribution) pairs
        if(outgoingLinks.size() > 0) {
            double massToSend = pageRank/outgoingLinks.size();
            for(String link: outgoingLinks) {
                result.add(new Tuple2<>(link, massToSend));
            }
        }

        return result.iterator();
    }
}
